### PR TITLE
Includes Artwork and Media Type for tracks.

### PR DESCRIPTION
**Pull request by codenift**
Includes the ability to retrieve artwork from audio files, generating thumbnails and data URI. Artwork is also added to the current API. 

Also includes Media Type for tracks. (Station ID, Show Intro, Show Outro, Jingles, Sweepers, Commercials, Promo, Interview, Shout Out, News, Voice Tracking).

### CODE REVIEW

File: airtime_mvc/application/models/airtime/CcFiles.php [hunk 1]
```diff
@@ -314,20 +315,19 @@ private static function validateFileArray(&$fileArray)
...
-		
-		$cueinSec = Application_Common_DateHelper::calculateLengthInSeconds($cuein);
-		$cueoutSec = Application_Common_DateHelper::calculateLengthInSeconds($cueout);
-		$lengthSec = bcsub($cueoutSec, $cueinSec, 6);
-		
-		$length = Application_Common_DateHelper::secondsToPlaylistTime($lengthSec);
-		
-		return $length;
-	}
+	  {
```
Comment (paddatrapper) [inline airtime_mvc/application/models/airtime/CcFiles.php]: This indentation seems to have increased the wrong way - should be less

File: airtime_mvc/application/upgrade/Upgrades.php [hunk 2]
```diff
@@ -494,6 +494,19 @@ public function getNewVersion() {
     }
 }
 
+class AirtimeUpgrader2517 extends AirtimeUpgrader
```
Comment (paddatrapper) [inline airtime_mvc/application/upgrade/Upgrades.php]: Has this been tested? Also should probably be a separate PR

File: airtime_mvc/application/views/helpers/VersionNotify.php [hunk 3]
```diff
@@ -36,6 +36,7 @@ public function versionNotify() {
         $config = Config::getConfig();
 
         // retrieve and validate current and latest versions,
+        //$current = $config['airtime_version'];
```
Comment (paddatrapper) [inline airtime_mvc/application/views/helpers/VersionNotify.php]: This can be deleted

File: airtime_mvc/application/views/scripts/embed/player.phtml [hunk 4]
```diff
@@ -339,6 +339,7 @@
                     } else {
                         $("ul.schedule_list").find("li").html(data.next.name);
                     }
+                    //console.log(data.current.metadata.artwork_data);
```
Comment (paddatrapper) [inline airtime_mvc/application/views/scripts/embed/player.phtml]: This can be deleted

File: airtime_mvc/public/js/airtime/library/library.js [hunk 5]
```diff
@@ -785,6 +794,7 @@ var AIRTIME = (function(AIRTIME) {
                 } else if (!aData.is_playlist) {
                     $(nRow).find("td.library_is_playlist").html('');
                 }
+                //console.log(aData);
```
Comment (paddatrapper) [inline airtime_mvc/public/js/airtime/library/library.js]: Please delete this line

File: airtime_mvc/public/js/airtime/login/login.js [hunk 6]
```diff
@@ -4,6 +4,6 @@ $(window).load(function() {
 
 $(document).ready(function() {
     $("#submit").click(function() {
-        $.cookie("airtime_locale", $("#locale").val(), {path: '/'});
+        Cookies.set('airtime_locale', $('#locale').val(), {path: '/'});
```
Comment (paddatrapper) [inline airtime_mvc/public/js/airtime/login/login.js]: Single quotes are inconsistent with the rest of the JavaScript

File: airtime_mvc/public/js/libs/angular.min.js [hunk 7]
```diff
@@ -1,293 +1,5 @@
-/*
```
Comment (paddatrapper) [inline airtime_mvc/public/js/libs/angular.min.js]: What is this change?

File: logo/144px.svg [hunk 8]
```diff
@@ -1,27 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (paddatrapper) [inline logo/144px.svg]: Please do not delete the logo directory and related files

File: airtime_mvc/application/common/FileDataHelper.php [hunk 9]
```diff
@@ -68,4 +68,369 @@ public static function getFileExtensionFromMime($mime)
...
+    /**
+     * Normalize string
+     *
+     * @param string $string
+     *
+     * @return string Normalized string
+     */
+    private static function normalizePath($string)
+    {
+        static $normal = array (
```
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/common/FileDataHelper.php]: Everything would work as it should. The only difference is that the file will be named without the character (eg. "Úsame.jpg" will then be "same.jpg") if we don't use normalize. I added it just to keep things more organized but can improved and even use the function to do the same for audio files since it currently saves files without those characters. So if we use it, "Marc Anthony Hits/same.mp3" would then be  "Marc_Anthony_Hits/Usame.mp3"

File: airtime_mvc/application/common/getid3/extension.cache.dbm.php [hunk 10]
```diff
@@ -0,0 +1,251 @@
+<?php
```
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/common/getid3/extension.cache.dbm.php]: It can. (https://github.com/JamesHeinrich/getID3) I tried doing it but must've done something wrong and since I needed it for my site I just went ahead and included it to LB. I understand. Keep in mind I'm still new to doing things with git and haven't had much time to dig in, actually my first time contributing so please bear with me.

File: airtime_mvc/application/controllers/ApiController.php [hunk 11]
```diff
@@ -567,15 +682,10 @@ public function stationLogoAction()
             $this->_helper->viewRenderer->setNoRender(true);
 
             $logo = Application_Model_Preference::GetStationLogo();
-            // if there's no logo, just die - redirects to a 404
```
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/controllers/ApiController.php]: Don't really remember doing so and must've been something I didn't notice pushing.

File: airtime_mvc/public/js/airtime/login/login.js (hunk 6 above)
Comment (codenift) [reply airtime_mvc/public/js/airtime/login/login.js]: Noted

File: airtime_mvc/public/js/libs/angular.min.js (hunk 7 above)
Comment (codenift) [reply airtime_mvc/public/js/libs/angular.min.js]: Removed //# sourceMappingURL=angular.min.js.map at the end of the file so that an error doesn't show up on some browsers.

File: airtime_mvc/application/models/Preference.php [hunk 12]
```diff
@@ -201,10 +201,11 @@ private static function getValue($key, $isUserValue = false, $forceDefault = fal
     public static function GetHeadTitle()
     {
         $title = self::getValue("station_name");
-        if (strlen($title) > 0)
-            $title .= " - ";
-
-        return $title.PRODUCT_NAME;
+        if ($title == null) {
```
[the remark this answers is not in the document]
Comment (hairmare) [reply airtime_mvc/application/models/Preference.php]: This branch already had merge conflicts when I created #871. I guess landing this might require a bit of effort since it seems to have quite some changes unrelated to the core feature being implemented in the branch.

File: airtime_mvc/application/common/FileDataHelper.php [hunk 13]
```diff
@@ -68,4 +68,369 @@ public static function getFileExtensionFromMime($mime)
...
+    /**
+     * Render image
+     * Used in API
+     *
+     * @param string $file
+     * @param string $size
+     */
+    public static function renderImage($file)
+    {
+          $im = @imagecreatefromjpeg($file);
```
Comment (hairmare) [inline airtime_mvc/application/common/FileDataHelper.php]: In my opinion the `@` operator should be purged from PHP itself. What kin of errors does this try to ignore and ca we handle and log them properly in some way?

File: airtime_mvc/application/common/FileDataHelper.php [hunk 14]
```diff
@@ -68,4 +68,369 @@ public static function getFileExtensionFromMime($mime)
...
+    /**
+     * Render Data URI
+     * Used in API
+     *
+     * @param string $dataFile
+     */
+    public static function renderDataURI($dataFile)
+    {
+          if($filecontent = file_get_contents($dataFile) !== false){
+               $image = @file_get_contents($dataFile);
```
Comment (hairmare) [inline airtime_mvc/application/common/FileDataHelper.php]: Can we handle potential errors and not ignore them using `@`?

File: airtime_mvc/application/common/FileDataHelper.php [hunk 15]
```diff
@@ -68,4 +68,369 @@ public static function getFileExtensionFromMime($mime)
...
+     * Convert image to Data URI
+     *
+     * @param string $orig_filename
+     * @param string $conv_filename
+     */
+    public static function imgToDataURI($orig_filename, $conv_filename)
+    {
+          $file = file_get_contents($orig_filename);
+          $Image = 'data:image/jpeg;charset=utf-8;base64,'.base64_encode($file);
+          $base64 = @$Image;
```
Comment (hairmare) [inline airtime_mvc/application/common/FileDataHelper.php]: This `@` only seems to ignore errors when assigning one string to another variable. Please remove it.

File: airtime_mvc/application/common/getid3/extension.cache.dbm.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (hairmare) [reply airtime_mvc/application/common/getid3/extension.cache.dbm.php]: I'll try helping on getting this landed for sure. Would doing and landing the composer change for you help in any way?

I could either do the change and push to you branch (if you have the allow maintainers edit checkbox set on your PR) or I could add the ID3 lib in a seperate PR and you could merge/rebase yours. Let me know what you would prefer.

File: airtime_mvc/application/common/FileDataHelper.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/common/FileDataHelper.php]: This was just my way of organizing files. Sure we can possibly save as unicode and decode, but I prefer having them already named if it's going to be a physical file and is something that could also be used in audio files seeing how LB also allows you to Download a file. If trying to download a song named "Úsame.mp3" right now, you're just going to get a file named 'same.mp3', so it will need be saved as unicode there as well and decoded right before a download. I just prefer making things a bit cleaner this way.

File: airtime_mvc/application/common/getid3/extension.cache.dbm.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/common/getid3/extension.cache.dbm.php]: Thanks. Either way would work. Yes, the allow maintainers edit is checked so you can do the change.

File: airtime_mvc/application/models/Preference.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/models/Preference.php]: Yes, that's my fault for not having seperate pull requests. I did the PR before it and suggested it in discourse, but I can just remove it and merge the one hairmare added since it's much cleaner.

File: airtime_mvc/application/models/airtime/CcFiles.php (hunk 1 above)
Comment (codenift) [reply airtime_mvc/application/models/airtime/CcFiles.php]: Oops. Didn't notice. I have to be more careful next time.

File: airtime_mvc/public/js/libs/angular.min.js (hunk 7 above)
Comment (paddatrapper) [reply airtime_mvc/public/js/libs/angular.min.js]: Ok. As this is an embedded upstream library, is that change specific to LT or has angular fixed it in a subsequent version?

File: airtime_mvc/application/common/FileDataHelper.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (paddatrapper) [reply airtime_mvc/application/common/FileDataHelper.php]: Sorry, my question was more this: If a file has a unicode char that is not on this list, what would break?

[the remark this answers is not in the document]
Comment (paddatrapper) [reply airtime_mvc/application/common/FileDataHelper.php]: > Sure we can possibly save as unicode and decode
What do you mean by this?

File: airtime_mvc/public/js/libs/angular.min.js (hunk 7 above)
Comment (codenift) [reply airtime_mvc/public/js/libs/angular.min.js]: Just on LT.

File: airtime_mvc/application/common/FileDataHelper.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (codenift) [reply airtime_mvc/application/common/FileDataHelper.php]: Nothing would break. Once it saves into the db, that's what's going to be used to from now on.

File: airtime_mvc/public/js/libs/angular.min.js (hunk 7 above)
Comment (paddatrapper) [reply airtime_mvc/public/js/libs/angular.min.js]: What is the error? Would a rebase on a newer version of Angular fix the problem?

File: airtime_mvc/application/common/getid3/extension.cache.dbm.php (hunk 10 above)
[the remark this answers is not in the document]
Comment (hairmare) [reply airtime_mvc/application/common/getid3/extension.cache.dbm.php]: Before I get to it, please bear in mind that you will need to update your local working copy once any change I do is pushed (ie. by running `git pull`).

We'll probably want to squash a few of your edits when we merge this but I guess we can wait until everything looks good before we squash.

**Comment by codenift** [comment]
Yes, that’s fine with me. Sorry I haven’t had much time to work on it this
week but will get to it as soon as I can.

**Comment by codenift** [comment]
I'm closing this PR to start all over with a clean fork since I have a better understanding now but don't want to drive myself crazy since I don't really have much time to study Github. I made a backup to compare all the changes with the main LB, and also made a PDF with the info on this page just incase. I'll be separating into different PRs.
